### ISSUE TITLE
"failed to verify certificate: x509: certificate signed by unknown authority" on k3s installation

### ISSUE THREAD

**Issue by latchmihay** [clastix/capsule-proxy#291]
# Bug description

capsule-proxy reports `tls: failed to verify certificate: x509: certificate signed by unknown authority`

# How to reproduce

1. Run a k3s cluster
2. Install Capsule
3. Install capsule-proxy
4. Generate an user `create-user.sh alice test`
5. Modify user's kubeconfig over to capsule-proxy port
6. export kubeconfig ^
7. execute a kubectl command

# Logs

### from capsule-proxy
```console
2023/05/11 20:37:29 http: TLS handshake error from 127.0.0.1:55596: tls: failed to verify certificate: x509: certificate signed by unknown authority
```

#### From `kubectl`:
```console
# kubectl get ns -v 9
I0511 13:46:58.000181 2911297 loader.go:374] Config loaded from file:  alice-test.capsule
I0511 13:46:58.003090 2911297 cert_rotation.go:137] Starting client certificate rotation controller
I0511 13:46:58.003172 2911297 round_trippers.go:466] curl -v -XGET  -H "User-Agent: kubectl/v1.24.13+k3s1 (linux/amd64) kubernetes/3f79b28" -H "Accept: application/json, */*" 'https://127.0.0.1:9001/api?timeout=32s'
I0511 13:46:58.004096 2911297 round_trippers.go:510] HTTP Trace: Dial to tcp:127.0.0.1:9001 succeed
I0511 13:46:58.036885 2911297 round_trippers.go:553] GET https://127.0.0.1:9001/api?timeout=32s  in 33 milliseconds
I0511 13:46:58.037014 2911297 round_trippers.go:570] HTTP Statistics: DNSLookup 0 ms Dial 0 ms TLSHandshake 29 ms Duration 33 ms
I0511 13:46:58.037111 2911297 round_trippers.go:577] Response Headers:
I0511 13:46:58.037422 2911297 cached_discovery.go:119] skipped caching discovery info due to Get "https://127.0.0.1:9001/api?timeout=32s": remote error: tls: bad certificate
```

# Additional context

- Capsule-Proxy version: v0.4.2
- Helm Chart version: capsule-proxy:v0.4.3
- Kubernetes version: 
```console
# kubectl version
WARNING: This version information is deprecated and will be replaced with the output from kubectl version --short.  Use --output=yaml|json to get the full version.
Client Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.13+k3s1", GitCommit:"3f79b289229dc0452c5cde26cccfa069805c5610", GitTreeState:"clean", BuildDate:"2023-04-20T01:39:02Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
Kustomize Version: v4.5.4
Server Version: version.Info{Major:"1", Minor:"24", GitVersion:"v1.24.13+k3s1", GitCommit:"3f79b289229dc0452c5cde26cccfa069805c5610", GitTreeState:"clean", BuildDate:"2023-04-20T01:39:02Z", GoVersion:"go1.19.8", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by prometherion** [comment]
This was an intriguing issue and I had the chance to learn more about how `k3s` managed certificates.

tl;dr; I was able to use client certificates based authentication on `k3s` using the following script.

```bash
#!/usr/bin/env bash

# This script generates a kubeconfig backed by client certificate
# signed by the k3s-server CA.
# It requires cluster admin permission.
#
# e.g.: ./create-user.sh alice oil
#       where `oil` is the Tenant and `alice` the owner

# Exit immediately if a command exits with a non-zero status.
set -e

function check_command() {
    local command=$1

    if ! command -v $command &> /dev/null; then
        echo "Error: ${command} not found"
        exit 1
    fi
}

check_command k3s
check_command openssl

USER=$1
TENANT=$2
GROUP=$3
CAPSULE_PROXY_ADDRESS=$4
CAPSULE_PROXY_CA_PATH=$5

if [[ -z ${USER} ]]; then
    echo "User has not been specified!"
    exit 1
fi

if [[ -z ${TENANT} ]]; then
    echo "Tenant has not been specified!"
    exit 1
fi

if [[ -z ${GROUP} ]]; then
    GROUP=capsule.clastix.io
fi

if [[ -z ${CAPSULE_PROXY_ADDRESS} ]]; then
    echo "Capsule Proxy address has not been specified!"
    exit 1
fi

if [[ -z ${CAPSULE_PROXY_CA_PATH} ]]; then
    echo "Capsule Proxy CA path has not been specified!"
    exit 1
fi

TMPDIR=$(mktemp -d)
echo "creating certs in TMPDIR ${TMPDIR} "

MERGED_GROUPS=$(echo "/O=$GROUP" | sed "s/,/\/O=/g")
echo "merging groups ${MERGED_GROUPS}"

K3S_FOLDER="/var/lib/rancher/k3s/server/tls"
CSR_PATH="${TMPDIR}/${USER}-${TENANT}.csr"
KEY_PATH="${TMPDIR}/${USER}-${TENANT}.key"
CRT_PATH="${TMPDIR}/${USER}-${TENANT}.crt"

# Generate cert and key based on the k3s client CA
openssl genrsa -out "${KEY_PATH}" 4096
openssl req -new -key "${KEY_PATH}" -out "${CSR_PATH}" -subj "/CN=${USER}${MERGED_GROUPS}"
openssl x509 -req -in "${CSR_PATH}" -CA "$K3S_FOLDER/server-ca.crt" -CAkey "$K3S_FOLDER/server-ca.key" -CAcreateserial -out "${CRT_PATH}" -days 36

# Create the kubeconfig file
CONTEXT=$(k3s kubectl config current-context)
CLUSTER=$(k3s kubectl config view -o jsonpath="{.contexts[?(@.name == \"$CONTEXT\"})].context.cluster}")

cat > ${USER}-${TENANT}.kubeconfig <<EOF
apiVersion: v1
clusters:
- cluster:
    certificate-authority-data: $(cat ${CAPSULE_PROXY_CA_PATH} | base64 | tr -d '\n')
    server: ${CAPSULE_PROXY_ADDRESS}
  name: ${CLUSTER}
contexts:
- context:
    cluster: ${CLUSTER}
    user: ${USER}
  name: ${USER}-${TENANT}
current-context: ${USER}-${TENANT}
kind: Config
preferences: {}
users:
- name: ${USER}
  user:
    client-certificate-data: $(cat ${CRT_PATH} | base64 | tr -d '\n')
    client-key-data: $(cat ${KEY_PATH} | base64 | tr -d '\n')
EOF

echo "kubeconfig file is:" ${USER}-${TENANT}.kubeconfig
echo "to use it as" ${USER} "export KUBECONFIG="${USER}-${TENANT}.kubeconfig

rm -rf "${TMPDIR}"
```

It seems `k3s` has two CAs: one for the server, and another for the client.

Upon the client-certificate verification, the generated tenant owner certificate was signed by the wrong CA (the client one) rather than the server one that is then used by `capsule-proxy`.

The provided script is gonna work if and only if it targets the `capsule-proxy`.

I was able to understand this thanks to the discussion that originated from k3s-io/k3s#684: however, I would suggest relying on bearer tokens without messing up with certs.

## Smoke test

```
$: k3s kubectl get tnt
NAME    STATE    NAMESPACE QUOTA   NAMESPACE COUNT   NODE SELECTOR   AGE
latch   Active                     1                                 61s

$ sudo ./script.sh alice latch capsule.clastix.io https://127.0.0.1:9001 /home/prometherion/.local/share/mkcert/rootCA.pem
creating certs in TMPDIR /tmp/tmp.kHHpjS8f2Q 
merging groups /O=capsule.clastix.io
Certificate request self-signature ok
subject=CN = alice, O = capsule.clastix.io
kubeconfig file is: alice-latch.kubeconfig
to use it as alice export KUBECONFIG=alice-latch.kubeconfig

$: KUBECONFIG=alice-latch.kubeconfig kubectl get ns
No resources found

$: KUBECONFIG=alice-latch.kubeconfig kubectl create ns latch-dev
namespace/latch-dev created

$: KUBECONFIG=alice-latch.kubeconfig kubectl get ns
NAME        STATUS   AGE
latch-dev   Active   1s
```

Let me know if it's good enough to close the issue.

**Comment by latchmihay** [comment]
Confirmed and works as expected, thank you for getting to the bottom of this!
